### PR TITLE
transport: replace isItem with a marker method

### PR DESCRIPTION
**Pull request by tamird**
Also changes the receivers to pointers because:
- Implementing an interface using value receivers causes both pointers
and values to implement that interface; implementing an interface using
pointer receivers causes only pointers to implement the interface,
thereby providing better type safety.
- Wrapping any value other than an empty struct in an interface causes
the value to be heap-allocated; no additional allocations are therefore
caused by this change.

### CODE REVIEW

File: transport/transport.go [hunk 1]
```diff
@@ -63,13 +63,11 @@ type recvMsg struct {
...
 }
 
-func (recvMsg) isItem() bool {
-	return true
-}
+func (*recvMsg) isItem() {}
 
 // All items in an out of a recvBuffer should be the same type.
 type item interface {
-	isItem() bool
```
Comment (dsymonds) [inline transport/transport.go]: The benefit of an isFoo method in an interface like this returning a bool is to permit wrapper types to implement this method correctly.

**Comment by dsymonds** [comment]
What you say is true about the method sets, but does that really matter for an unexported interface, with only unexported types implementing that interface?

File: transport/transport.go (hunk 1 above)
Comment (tamird) [inline transport/transport.go]: I don't understand this comment. This is an unexported interface - what wrapper types are you envisioning?

Comment (dsymonds) [inline transport/transport.go]: I don't know. I didn't write this code originally. I'm just saying that this could be a reason why the `bool` is here.

You didn't mention this change in the commit message. Making unnecessary or unrelated changes at the same time as a real change makes it harder to review. Please try to avoid that.

Comment (tamird) [inline transport/transport.go]: I did mention it - "isItem is a marker method". Maybe that didn't come across.

**Comment by tamird** [comment]
Sure, it matters. It makes it impossible to make a certain class of mistake since it will now be caught by the compiler. It's just tidying up, but that's worthwhile.

File: transport/control.go [hunk 2]
```diff
@@ -56,43 +56,33 @@ type windowUpdate struct {
...
 	code     http2.ErrCode
 }
 
-func (resetStream) isItem() bool {
-	return true
-}
+func (*resetStream) isItem() {}
 
 type flushIO struct {
 }
```
Comment (iamqizhao) [inline transport/control.go]: this is an empty struct so that the value receiver does have advantage here.

File: transport/transport.go [hunk 3]
```diff
@@ -63,13 +63,11 @@ type recvMsg struct {
...
+	isItem()
```
Comment (iamqizhao) [inline transport/transport.go]: I do not see the point to make this change. What is the benefit here?

File: transport/control.go (hunk 2 above)
Comment (tamird) [inline transport/control.go]: it has an advantage if you actually use it as a value - all current uses are using pointers, and that's the problem with value receivers in general; they permit both.

File: transport/transport.go (hunk 3 above)
Comment (tamird) [inline transport/transport.go]: it is just a simplification.

**Comment by tamird** [comment]
@iamqizhao anything more I can answer here?

File: transport/control.go (hunk 2 above)
Comment (iamqizhao) [inline transport/control.go]: I am still not convinced we should do this. According to golang faq,
"For types such as basic types, slices, and small structs, a value receiver is very cheap so unless the semantics of the method requires a pointer, a value receiver is efficient and clear."

Actually I am thinking we should change the caller's code instead.

Comment (tamird) [inline transport/control.go]: That section of the FAQ is completely irrelevant. This method is never
called, this is about interface-induced allocation and type safety.

File: transport/transport.go (hunk 1 above)
Comment (xiang90) [inline transport/transport.go]: the method naming is probably confusing... I would expect isItem to return true, or false. I would not expect this method to exist as a marker item.

File: transport/control.go (hunk 2 above)
Comment (iamqizhao) [inline transport/control.go]: It is relevant because we are constructing a samll struct here and try to decide it should be value receiver or pointer receiver now. How a struct is defined should NOT depend on its caller's behavior.

File: transport/transport.go (hunk 3 above)
Comment (iamqizhao) [inline transport/transport.go]: I feel non-sense to make this change. Can you revert this part?

File: transport/transport.go (hunk 1 above)
Comment (tamird) [inline transport/transport.go]: @xiang90 do you think it should just be `item()`?

Comment (xiang90) [inline transport/transport.go]: @tamird If this is a marker method, I feel so. There are some code in stdlib using similar approach (https://golang.org/src/go/ast/ast.go: type Expr interface is an example)

Comment (tamird) [inline transport/transport.go]: done

File: transport/transport.go [hunk 4]
```diff
@@ -63,13 +63,11 @@ type recvMsg struct {
...
 
-func (recvMsg) isItem() bool {
-	return true
-}
+func (*recvMsg) Item() {}
 
 // All items in an out of a recvBuffer should be the same type.
 type item interface {
-	isItem() bool
+	Item()
```
Comment (iamqizhao) [inline transport/transport.go]: unexported

Comment (tamird) [inline transport/transport.go]: done
